File: test_src/fortran/hierarchical_parallelism/reduction/real/target_simd.f90


program target_simd.f90


    

    implicit none
  
    INTEGER :: L = 5
    INTEGER :: i
    
    REAL :: COUNTER =  0   

    
    
     
    
    !$OMP TARGET SIMD   REDUCTION(+:COUNTER)   MAP(TOFROM: COUNTER) 


    DO i = 1 , L 


    

    


counter = counter +  1.  

 
     

    END DO

    !$OMP END TARGET SIMD
    

    IF  ( ( ABS(COUNTER - L) ) > 10*EPSILON( COUNTER   ) ) THEN
        write(*,*)  'Expected L Got', COUNTER
        call exit(1)
    ENDIF

end program target_simd.f90

File: src/test_src/cpp/hierarchical_parallelism/reduction-float/target__simd.cpp
#include <iostream>
#include <cstdlib>
bool almost_equal(float x, float gold, float tol) {
        return gold * (1-tol) <= x && x <= gold * (1 + tol);
}
void test_target__simd(){
 const int N_i = 64;
 float counter{};
#pragma omp target map(tofrom: counter) 
#pragma omp simd reduction(+: counter)
    for (int i = 0 ; i < N_i ; i++ )
    {
counter += float { 1.0f };
    }
if ( !almost_equal(counter,float { N_i }, 0.1)  ) {
    std::cerr << "Expected: " << N_i << " Got: " << counter << std::endl;
    std::exit(112);
}
}
int main()
{
    test_target__simd();
}
